pytorch/pytorch_ddp_1machine.py

```diff
@@ -0,0 +1,216 @@
+PyTorch distributed data parallel and FashionMNIST.
+This example is dedicated to a use case of using one machine with multiple GPUs
+where no MPI is needed.
+We optimize the neural network architecture as well as the optimizer configuration.
+As it is too time consuming to use the whole FashionMNIST dataset, we
+here use a small subset of it.
+
+You can execute this example with a command as follows:
+    $ python pytorch_ddp_1machine.py
+
+"""
+
+from functools import partial
+import os
+
+import optuna
+from optuna.trial import TrialState
+import torch
+import torch.distributed as dist
+import torch.multiprocessing as mp
+import torch.nn as nn
+import torch.nn.functional as F
+from torch.nn.parallel import DistributedDataParallel as DDP
+import torch.optim as optim
+import torch.utils.data
+from torchvision import datasets
+from torchvision import transforms
+
+
+# Arbitrary GPUs can be selected
+DEVICE_IDS = [0, 1]
+BATCHSIZE = 128
+CLASSES = 10
+DIR = os.getcwd()
+EPOCHS = 10
+N_TRAIN_EXAMPLES = BATCHSIZE * 30
+N_VALID_EXAMPLES = BATCHSIZE * 10
+N_TRIALS = 20
+
+
+def define_model(trial):
+    # We optimize the number of layers, hidden units and dropout ratio in each layer.
+    n_layers = trial.suggest_int("n_layers", 1, 3)
+    layers = nn.ModuleList()
+
+    in_features = 28 * 28
+    for i in range(n_layers):
+        out_features = trial.suggest_int("n_units_l{}".format(i), 4, 128)
+        layers.append(nn.Linear(in_features, out_features))
+        layers.append(nn.ReLU())
+        p = trial.suggest_float("dropout_l{}".format(i), 0.2, 0.5)
+        layers.append(nn.Dropout(p))
+
+        in_features = out_features
+    layers.append(nn.Linear(in_features, CLASSES))
+    layers.append(nn.LogSoftmax(dim=1))
+
+    return nn.Sequential(*layers)
+
+
+def get_mnist():
+    # Load FashionMNIST dataset.
+    train_dataset = datasets.FashionMNIST(DIR, train=True, transform=transforms.ToTensor())
+    train_dataset = torch.utils.data.Subset(train_dataset, indices=range(N_TRAIN_EXAMPLES))
+    train_sampler = torch.utils.data.distributed.DistributedSampler(dataset=train_dataset)
+
+    valid_dataset = datasets.FashionMNIST(DIR, train=False, transform=transforms.ToTensor())
+    valid_dataset = torch.utils.data.Subset(valid_dataset, indices=range(N_VALID_EXAMPLES))
+    valid_sampler = torch.utils.data.distributed.DistributedSampler(
+        dataset=valid_dataset, shuffle=False
+    )
+
+    train_loader = torch.utils.data.DataLoader(
+        train_dataset,
+        sampler=train_sampler,
+        batch_size=BATCHSIZE,
+        shuffle=False,
+    )
+    valid_loader = torch.utils.data.DataLoader(
+        valid_dataset,
+        sampler=valid_sampler,
+        batch_size=BATCHSIZE,
+        shuffle=False,
+    )
+
+    return train_loader, valid_loader, train_sampler, valid_sampler
+
+
+def objective(single_trial, device_id):
+    device = torch.device("cuda", device_id)
```

Comment:
Do we need this variable? I thought `rank` was enough when I implemented similar example
https://gist.github.com/nzw0301/9ba3c837a02539e8bde194f4f2c9d0e5.

Reply:
The code still runs without converting `device_id` to `device` but I'm not sure how did your example get the variable `rank` in 
```
def objective(single_trial):
    trial = optuna.integration.TorchDistributedTrial(single_trial, rank)
```
I can modify the code so that it only uses `device_id`, but I think `objective` still needs an argument for device_id unless I'm missing something.

Reply:
Hmm, indeed. Thank you for your comment. As you did, `partial` or `lambda` might be necessary.

Reply:
> def objective(single_trial):
>    trial = optuna.integration.TorchDistributedTrial(single_trial, rank)

The new `TorchDistributedTrial` in optuna==v3.1.0 will not require `rank` or `device`.
It creates a process group with gloo backend only for Optuna's communication. 
So, we may be able to update this part after the release of optuna v3.1.0.
https://github.com/optuna/optuna/blob/25a135c26662e5e26f88d7266eb2922703bf72dd/optuna/integration/pytorch_distributed.py#L119

Reply:
Thank you @toshihikoyanase. Indeed having `rank` argument yields an error with optuna v3.1.0, but without this argument, optuna < 3.1.0 will raise an error (i.e. backward incompatible).  I can push a commit to remove `rank` argument, and perhaps with a comment in the beginning of the code to let users know that this example only works with optuna >= 3.1.0. 
Also, do we need to specify optuna version in the [requirements.txt](https://github.com/li-li-github/optuna-examples/blob/main/pytorch/requirements.txt)?